### DISCUSSION TITLE
Is this the expected behavior of pack and unpack ID?

### DISCUSSION THREAD

**Discussion by aVersionOfReality** [bnpr/Malt#563] [General]
Pack/Unpack ID is not behaving the way I thought it would. I thought that whatever channels went in, would be the same when they come out. But it seems that the input Blue and Alpha influence the Red and Green, and nothing comes out of them in the main pass?

Is this a bug, or am I misunderstanding how this 8bit packing works?

https://github.com/bnpr/Malt/assets/36803956/14e4a9c2-8bb0-4aff-b9ee-ec95bae5f1e1

**Comment by pragma37** [comment]
No, this is not expected to work (note that there's not really an "Unpack ID" node).
Packing a color to an ID converts it to a 16bit value, so it can't store the full 32 bit color.
https://github.com/bnpr/Malt/blob/f46cc9484f073cbeeefa86a1c72b811a1dd15ef1/Malt/Pipelines/NPR_Pipeline/Shaders/NPR_Pipeline/NPR_Mesh.glsl#L192

The Id node values can be used to compare them to other values stored in the `IN_ID` texture, but can't be unpacked back to its original value.

EDIT: It will work if the B and A channels are set to 0, but I wouldn't rely on that, since this is an implementation detail that may change in the future.

**Reply by aVersionOfReality** [reply]
Ah, alright. But it does seem to work for a single float passed through R? Or might that change too?

**Reply by pragma37** [reply]
That's what I meant, if B and A channels are set to 0, the R and G channels can be unpacked back to their original values.
But don't rely on that since that could change too.
